From: Gleb Pesin <dormancygrace@gmail.com>
Date: Wed, 29 Jul 2026 21:00:00 +0000
Subject: [PATCH] mt7620: use long calls in affected objects

MT7620 can sporadically execute direct J-type calls with an incorrect high
address nibble.  Linux 6.18 code generation and layout exposed failures in
the entropy bootstrap, SPI core and IPv6 UDP code.

Build only the objects containing the observed failure functions with long
calls.  Applying -mlong-calls to the whole kernel was tested on MT7620N
hardware and prevents PDMA from processing its first transmit descriptor.
The targeted workaround boots and passes sustained bidirectional traffic
while retaining the known-good frame-engine code generation.

Signed-off-by: Gleb Pesin <dormancygrace@gmail.com>
---
 drivers/char/Makefile | 4 ++++
 drivers/spi/Makefile  | 4 ++++
 net/ipv6/Makefile     | 4 ++++
 3 files changed, 12 insertions(+)

--- a/drivers/char/Makefile
+++ b/drivers/char/Makefile
@@ -3,6 +3,10 @@
 # Makefile for the kernel character device drivers.
 #
 
+ifdef CONFIG_NET_RALINK_MT7620
+CFLAGS_random.o			+= -mlong-calls
+endif
+
 obj-y				+= mem.o random.o
 obj-$(CONFIG_TTY_PRINTK)	+= ttyprintk.o
 obj-y				+= misc.o
--- a/drivers/spi/Makefile
+++ b/drivers/spi/Makefile
@@ -5,6 +5,10 @@
 
 ccflags-$(CONFIG_SPI_DEBUG) := -DDEBUG
 
+ifdef CONFIG_NET_RALINK_MT7620
+CFLAGS_spi.o				+= -mlong-calls
+endif
+
 # small core, mostly translating board-specific
 # config declarations into driver model code
 obj-$(CONFIG_SPI_MASTER)		+= spi.o
--- a/net/ipv6/Makefile
+++ b/net/ipv6/Makefile
@@ -5,6 +5,10 @@
 
 obj-$(CONFIG_IPV6) += ipv6.o
 
+ifdef CONFIG_NET_RALINK_MT7620
+CFLAGS_udp.o += -mlong-calls
+endif
+
 ipv6-y :=	af_inet6.o anycast.o ip6_output.o ip6_input.o addrconf.o \
 		addrlabel.o \
 		route.o ip6_fib.o ipv6_sockglue.o ndisc.o udp.o udplite.o \
